README.md
Description
===========

*Please send me feedback and comments to carlitos.kyo@gmail.com*

Based on djangorecipe and code from setuptools used.

This buildout recipe can be used to create a setup for Django. It will
automatically download Django and install it in the buildout's
sandbox. You can use either a release version of Django or a
subversion checkout (by using `trunk` instead of a version number).

The directory structure is based on: http://django.es/blog/convenciones-proyecto-django/

You can see an example of how to use the recipe below::

  [buildout]
  parts = satchmo django
  eggs = ipython

  [satchmo]
  recipe = gocept.download
  url = http://www.satchmoproject.com/snapshots/satchmo-0.6.tar.gz
  md5sum = 659a4845c1c731be5cfe29bfcc5d14b1

  [django]
  recipe = djbuild
  version = trunk
  settings = development
  eggs = ${buildout:eggs}
  extra-paths =
    ${satchmo:location}
  project = dummyshop


Supported options
=================

The recipe supports the following options.

apps 
  projects that can be installed using pypi or compressed files. No handle
  dependencies do it by hand using buildout, the decision was taken for these reasons:
  
  * if dependency is a django app this should be declared into this option to install
    it into the extarnal-apps directory or it should be omited if the dependency
    was customized and it is on local-apps directory
    
  * if dependency is not a django app this should be declared into eggs option.
  
  To delete an application should be by hand.

project
  This option sets the name for your project. The recipe will create a
  basic structure if the project is not already there.
  
external-apps
  This option sets the directory where external reusable apps goes. Which do not
  be installed as an egg or if you don't want install it as an egg.
  
local-apps
  This option sets the directory where local reusable apps goes, usually
  put the company name for this directory, and customized apps.

projectegg
  Use this instead of the project option when you want to use an egg
  as the project. This disables the generation of the project
  structure.

python
  This option can be used to specify a specific Python version which can be a
  different version from the one used to run the buildout.

version
  The version argument can accept a few different types of
  arguments. You can specify `trunk`. In this case it will do a
  checkout of the Django trunk. Another option is to specify a release
  number like `0.96.2`. This will download the release
  tarball. Finally you can specify a full svn url (including the
  revision number). An example of this would be
  `http://code.djangoproject.com/svn/django/branches/newforms-admin@7833`.

settings
  You can set the name of the settings file which is to be used with
  this option. This is useful if you want to have a different
  production setup from your development setup. It defaults to
  `development`.

download-cache
  Set this to a folder somewhere on you system to speed up
  installation. The recipe will use this folder as a cache for a
  downloaded version of Django.

extra-paths
  All paths specified here will be used to extend the default Python
  path for the `bin/*` scripts.

pth-files
  Adds paths found from a site `.pth` file to the extra-paths.
  Useful for things like Pinax which maintains its own external_libs dir.

control-script
  The name of the script created in the bin folder. This script is the
  equivalent of the `manage.py` Django normally creates. By default it
  uses the name of the section (the part between the `[ ]`).

test
  If you want a script in the bin folder to run all the tests for a
  specific set of apps this is the option you would use. Set this to
  the list of app labels which you want to be tested.

testrunner
  This is the name of the testrunner which will be created. It
  defaults to `test`.
  
find-links
  used to install apps

All following options only have effect when the project specified by
the project option has not been created already, on the setting file 
especified.


FCGI specific settings
======================

Options for FCGI can be set within a settings file (`settings.py`). The options
is `FCGI_OPTIONS`. It should be set to a dictionary. The part below is an
example::

  FCGI_OPTIONS = {
      'method': 'threaded',
      'daemonize': 'false',
  }


Another example
===============

The next example shows you how to use some more of the options::

  [buildout]
  parts = django extras
  eggs =
    hashlib

  [extras]
  recipe = iw.recipe.subversion
  urls =
    http://django-command-extensions.googlecode.com/svn/trunk/ django-command-extensions
    http://django-mptt.googlecode.com/svn/trunk/ django-mptt

  [django]
  recipe = djbuild
  version = trunk
  settings = development
  project = exampleproject
  wsgi = true
  eggs =
    ${buildout:eggs}
  test =
    someapp
    anotherapp

Example using .pth files
========================

Pinax uses a .pth file to add a bunch of libraries to its path; we can
specify it's directory to get the libraries it specified added to our
path::

  [buildout]
  parts	= PIL
	  svncode
	  myproject

  [PIL]
  recipe	= zc.recipe.egg:custom
  egg		= PIL
  find-links	= http://dist.repoze.org/

  [svncode]
  recipe	= iw.recipe.subversion
  urls		= http://svn.pinaxproject.com/pinax/tags/0.5.1rc1	pinax

  [myproject]
  recipe	= djbuild
  version	= 1.0.2
  eggs		= PIL
  project	= myproject
  settings	= settings
  extra-paths	= ${buildout:directory}/myproject/apps
		  ${svncode:location}/pinax/apps/external_apps
		  ${svncode:location}/pinax/apps/local_apps
  pth-files	= ${svncode:location}/pinax/libs/external_libs
  wsgi		= true

Above, we use stock Pinax for pth-files and extra-paths paths for
apps, and our own project for the path that will be found first in the
list.  Note that we expect our project to be checked out (e.g., by
svn:external) directly under this directory in to 'myproject'.

Example with a different Python version
=======================================

To use a different Python version from the one that ran buildout in the
generated script use something like::

  [buildout]
  parts	= myproject

  [special-python]
  executable = /some/special/python

  [myproject]
  recipe	= djbuild
  version	= 1.0.2
  project	= myproject
  python	= special-python


Example configuration for mod_wsgi
==================================

If you want to deploy a project using mod_wsgi you could use this
example as a starting point::

  <Directory /path/to/buildout>
         Order deny,allow
         Allow from all
  </Directory>
  <VirtualHost 1.2.3.4:80>
         ServerName      my.rocking.server
         CustomLog       /var/log/apache2/my.rocking.server/access.log combined
         ErrorLog        /var/log/apache2/my.rocking.server/error.log
         WSGIScriptAlias / /path/to/buildout/bin/django.wsgi
  </VirtualHost>


CHANGES.txt
Changes
=======
Adding automatic install projects into external-apps from 
pypi or custom url

09-Oct-2010: deleting install call where installed mtweet on dist_installer.py
24-Oct-2010: fixing bug adding src directory into sys.path to find module settings (django project)
24-Oct-2010: fixing bug adding src directory into extra-path to generete script to find project into src directory
25-Oct-2010: fixing bug find-link option now into the correct place to install
25-Oct-2010: when creating the project creating templates dir
25-Oct-2010: adding base templates base.html, 404.html, 500.html
25-Oct-2010: updating templates
05-Nov-2010: adding urls into projects and logs directory
07-Nov-2010: fixing sdist package
07-Nov-2010: fixing settings template to load correct url, fixing url into producction
13-Nov-2010: fixing code that crash when changing directory names into local-apps external-apps
25-Feb-2011: moving project/test to project/testing
28-Mar-2011: updating to use django1.3 upload files moved to media and django static is as before #read django-staticfiles
28-Mar-2011: creating src dir if not exists
29-Mar-2011: fixing template urls, it didn't use global url not it does.
08-Apr-2011: fixing template urls, know serves uploaded files in development mode.
07-Oct-2011: new template for base.html using html5


src/djbuild/djbuild.py
import logging, os, zc.buildout
import sys
import os

from installer import Installer

class DjBuild(object):

    def __init__(self, buildout, name, options):
        self.log = logging.getLogger(name)
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)

        self.buildout, self.name, self.options = buildout, name, options
        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)
        options['bin-directory'] = buildout['buildout']['bin-directory']

        options.setdefault('project', 'example.com')
        options.setdefault('external-apps', 'apps')
        options.setdefault('local-apps', 'company')
        options.setdefault('settings', 'development')

        options.setdefault('urlconf', options['project'] + '.urls')
        # Set this so the rest of the recipe can expect the values to be
        # there. We need to make sure that both pythonpath and extra-paths are
        # set for BBB.
        if 'extra-paths' in options:
            options['pythonpath'] = options['extra-paths']
        else:
            options.setdefault('extra-paths', options.get('pythonpath', ''))
            
        # Usefull when using archived versions
        buildout['buildout'].setdefault(
            'download-cache',
            os.path.join(buildout['buildout']['directory'],
                         'downloads'))

        # only try to download stuff if we aren't asked to install from cache
        self.install_from_cache = self.buildout['buildout'].get(
            'install-from-cache', '').strip() == 'true'
        
        self.__installer = Installer(self.options, self.buildout, self.log, self.name)


    def install(self):
        print '\n------------------ Installing DjBuild ------------------\n'
        version = self.options['version']
        location = self.options['location']
        base_dir = self.buildout['buildout']['directory']
        project_dir = os.path.join(base_dir, 'src')
        download_dir = self.buildout['buildout']['download-cache']
        
        # updating sys.path to find django settings
        sys.path.insert(0, project_dir)
        
        extra_path = self.__installer.get_extra_paths()
        requirements, ws = self.egg.working_set(['djbuild'])
        
        self.__installer.verify_or_create_download_dir(download_dir)
        self.__installer.remove_pre_existing_installation(location)
        
        self.__installer.install_django(version, download_dir, location, self.install_from_cache)
        self.__installer.install_recipe(location)
        self.__installer.install_project(project_dir, self.options['project'])
        
        script_paths = self.__installer.install_scripts(extra_path, ws)
        
        print '\n------------------ Installing DjBuild ------------------\n'

        return script_paths + [location]

    def update(self):
        pass
    


src/djbuild/templates.py
production_settings = '''from %(project)s.settings import *

# DEBUG = False # is better do this on global settings files, at root project directory

INSTALLED_APPS += (
    # Uncomment the next line to enable the admin:
    # 'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',
)

ROOT_URLCONF = '%(project)s.project.production.urls'
'''

production_urls = '''from django.conf.urls.defaults import patterns, include, url

from %(project)s.urls import urlpatterns

urlpatterns += patterns('',
    #(r'^pushinit/', include('myproject.urls')),
)
'''

development_settings = '''from %(project)s.settings import *

import os

DEBUG=True
TEMPLATE_DEBUG=DEBUG

# set this to '' for development it causes problems
# it is used to collect static files on production mode
# not necessary here
STATIC_ROOT = ''

# STATIC_ROOT path should be added in development mode to server those files
STATICFILES_DIRS += (
    os.path.join(SITE_ROOT, 'static'),
)

INSTALLED_APPS += (
    # Uncomment the next line to enable the admin:
    # 'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',
)

ROOT_URLCONF = '%(project)s.project.development.urls'
'''

development_urls = '''from django.conf import settings
from django.conf.urls.static import static
from django.conf.urls.defaults import patterns, include, url

from django.contrib.staticfiles.urls import staticfiles_urlpatterns

from %(project)s.urls import urlpatterns

urlpatterns += patterns('',
    #(r'^init/', include('myproject.urls')),
) + staticfiles_urlpatterns() + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
'''

testing_settings = '''
from %(project)s.settings import *
DEBUG=True
TEMPLATE_DEBUG=DEBUG

INSTALLED_APPS += (
    # Uncomment the next line to enable the admin:
    # 'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',
)

ROOT_URLCONF = '%(project)s.project.testing.urls'
'''

testing_urls = '''from django.conf.urls.defaults import patterns, include, url

from %(project)s.urls import urlpatterns

urlpatterns += patterns('',
    #(r'^init/', include('myproject.urls')),
)
'''

wsgi = '''#!/usr/bin/python

import os
import sys

# redirect sys.stdout to sys.stderr for bad libraries like geopy that uses
# print statements for optional import exceptions.
sys.stdout = sys.stderr

sys.path[0:0] = [
  %(path)s,
]

os.environ["DJANGO_SETTINGS_MODULE"] = '%(arguments)s'
from django.conf import settings

from django.core.handlers.wsgi import WSGIHandler
application = WSGIHandler()

'''

fcgi = ''' #!/usr/bin/python

import os
import sys

sys.path[0:0] = [
  %(path)s,
]

os.environ["DJANGO_SETTINGS_MODULE"] = '%(arguments)s'
from django.conf import settings

from django.core.servers.fastcgi import runfastcgi
FCGI_OPTIONS = getattr(settings, 'FCGI_OPTIONS', {})
runfastcgi(FCGI_OPTIONS)
'''

base_html = '''{% load i18n %}<!DOCTYPE html>

<!--[if IE 7 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie7"> <![endif]-->
<!--[if IE 8 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie8"> <![endif]-->
<!--[if IE 9 ]><html lang="{{ LANGUAGE_CODE }}" class="no-js ie9"> <![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html lang="{{ LANGUAGE_CODE }}" class="no-js"> <!--<![endif]-->

<head>

  <meta charset="utf-8">
  
  {% if meta_description %}<meta name="description" content="{{ meta_description }}">{% endif %}
  {% if meta_keywords %}<meta name="keywords" content="{{ meta_keywords }}">{% endif %}
  <meta name="author" content="linkux-it">
  <meta name="robots" content="index, follow">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <!-- favicon 16x16 -->
  <link rel="shortcut icon" href="/{{ STATIC_URL }}images/favicon.ico">
  <!-- apple touch icon 57x57 -->
  <link rel="apple-touch-icon" href="{{ STATIC_URL }}images/favicon.png">
  
  <script type="text/javascript" src="http://code.jquery.com/jquery.min.js"></script> 
  
  <!-- Main style sheet. Change version number in query string to force styles refresh -->
  <!-- Link element no longer needs type attribute -->
  <link rel="stylesheet" href="css/screen.css?v=1.0">

  <!-- Modernizr for feature detection of CSS3 and HTML5; must be placed in the "head" -->
  <!-- Script tag no longer needs type attribute -->
  <script src="js/modernizr-1.6.min.js"></script>

  <!-- Remove the script reference below if you're using Modernizr -->
  <!--[if lt IE 9]>
  <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->
  
  {% block extra_base_head %}{% endblock %}
  
  {% block extra_head %}{% endblock %}

  <title>HTML5 template{% block title %}{% if title %} | {{ title }}{% endif %}{% endblock %}</title>
  
</head>

<body>
  
  <header>
  </header>

  <footer>
  </footer>
  
</body>

</html>
'''

t_404_html = '''{% extends 'base.html' %}
{% load i18n %}
{% block container %}
{% trans 'Page Not Found' %}
{% endblock %}''' 

t_500_html = '''{% extends 'base.html' %}
{% load i18n %}
{% block container %}
{% trans 'Server error' %}
{% endblock %}'''


src/djbuild/installer.py
import os
import re
import shutil
import urllib2
import templates
import setuptools
import subprocess
import zc.recipe.egg

from zc.buildout import UserError
from dist_installer import install

class Installer:
    
    def __init__(self, options, buildout, log, name):
        self.options = options
        self.buildout = buildout
        self.log = log
        self.name = name
    
    def install_svn_version(self, version, download_dir, location,
                            install_from_cache):
        svn_url = self.version_to_svn(version)
        download_location = os.path.join(
            download_dir, 'django-' +
            self.version_to_download_suffix(version))
        if not install_from_cache:
            if os.path.exists(download_location):
                if self.svn_update(download_location, version):
                    raise UserError(
                        "Failed to update Django; %s. "
                        "Please check your internet connection." % (
                            download_location))
            else:
                self.log.info("Checking out Django from svn: %s" % svn_url)
                cmd = 'svn co %s %s' % (svn_url, download_location)
                if not self.buildout['buildout'].get('verbosity'):
                    cmd += ' -q'
                if self.command(cmd):
                    raise UserError("Failed to checkout Django. "
                                    "Please check your internet connection.")
        else:
            self.log.info("Installing Django from cache: " + download_location)

        shutil.copytree(download_location, location)
        return download_location


    def install_release(self, version, download_dir, tarball, destination):
        extraction_dir = os.path.join(download_dir, 'django-archive')
        setuptools.archive_util.unpack_archive(tarball, extraction_dir)
        # Lookup the resulting extraction dir instead of guessing it
        # (Django releases have a tendency not to be consistend here)
        untarred_dir = os.path.join(extraction_dir,
                                    os.listdir(extraction_dir)[0])
        shutil.move(untarred_dir, destination)
        shutil.rmtree(extraction_dir)
        return destination

    def get_release(self, version, download_dir):
        tarball = os.path.join(download_dir, 'django-%s.tar.gz' % version)

        # Only download when we don't yet have an archive
        if not os.path.exists(tarball):
            download_url = 'http://www.djangoproject.com/download/%s/tarball/'
            self.log.info("Downloading Django from: %s" % (
                    download_url % version))

            tarball_f = open(tarball, 'wb')
            f = urllib2.urlopen(download_url % version)
            tarball_f.write(f.read())
            tarball_f.close()
            f.close()
        return tarball
    
    def is_svn_url(self, version):
        # Search if there is http/https/svn or svn+[a tunnel identifier] in the
        # url or if the trunk marker is used, all indicating the use of svn
        svn_version_search = re.compile(
            r'^(http|https|svn|svn\+[a-zA-Z-_]+)://|^(trunk)$').search(version)
        return svn_version_search is not None

    def version_to_svn(self, version):
        if version == 'trunk':
            return 'http://code.djangoproject.com/svn/django/trunk/'
        else:
            return version

    def version_to_download_suffix(self, version):
        if version == 'trunk':
            return 'svn'
        return [p for p in version.split('/') if p][-1]

    def svn_update(self, path, version):
        command = 'svn up'
        revision_search = re.compile(r'@([0-9]*)$').search(
            self.options['version'])

        if revision_search is not None:
            command += ' -r ' + revision_search.group(1)
        self.log.info("Updating Django from svn")
        if not self.buildout['buildout'].get('verbosity'):
            command += ' -q'
        return self.command(command, cwd=path)

    def get_extra_paths(self):
        basic_dir = os.path.join(self.buildout['buildout']['directory'], 'src')
        local_apps_dir = os.path.join(basic_dir, self.options['project'], self.options['local-apps'])
        external_apps_dir = os.path.join(basic_dir, self.options['project'], self.options['external-apps'])
        extra_paths = [self.options['location'],
                       basic_dir, # now insert into extra path basic dir project
                       local_apps_dir, external_apps_dir]

        # Add libraries found by a site .pth files to our extra-paths.
        if 'pth-files' in self.options:
            import site
            for pth_file in self.options['pth-files'].splitlines():
                pth_libs = site.addsitedir(pth_file, set())
                if not pth_libs:
                    self.log.warning(
                        "No site *.pth libraries found for pth_file=%s" % (
                         pth_file,))
                else:
                    self.log.info("Adding *.pth libraries=%s" % pth_libs)
                    self.options['extra-paths'] += '\n' + '\n'.join(pth_libs)

        pythonpath = [p.replace('/', os.path.sep) for p in
                      self.options['extra-paths'].splitlines() if p.strip()]

        extra_paths.extend(pythonpath)
        
        return extra_paths
    
    def command(self, cmd, **kwargs):
        output = subprocess.PIPE
        if self.buildout['buildout'].get('verbosity'):
            output = None
        command = subprocess.Popen(
            cmd, shell=True, stdout=output, **kwargs)
        return command.wait()

    def create_file(self, file, template, options=None):
        f = open(file, 'w')
        if options is not None:
            f.write(template % options)
        else:
            f.write(template)
        f.close()
    
    def create_manage_script(self, extra_paths, ws):
        project = self.options.get('projectegg', self.options['project'])
        return zc.buildout.easy_install.scripts(
            [(self.options.get('control-script', self.name),
              'djbuild.manage', 'main')],
            ws, self.options['executable'], self.options['bin-directory'],
            extra_paths = extra_paths,
            arguments= "'%s.project.%s.settings'" % (project,
                                                     self.options['settings']))
    
    def create_test_runner(self, extra_paths, working_set):
        apps = self.options.get('test', '').split()
        # Only create the testrunner if the user requests it
        if apps:
            return zc.buildout.easy_install.scripts(
                [(self.options.get('testrunner', 'test'),
                  'djangorecipe.test', 'main')],
                working_set, self.options['executable'],
                self.options['bin-directory'],
                extra_paths = extra_paths,
                arguments= "'%s.project.%s.settings', %s" % (
                    self.options['project'],
                    self.options['settings'],
                    ', '.join(["'%s'" % app for app in apps])))
        else:
            return []
        
    def make_scripts(self, extra_paths, ws):
        scripts = []
        _script_template = zc.buildout.easy_install.script_template
        for protocol in ('wsgi', 'fcgi'):
            zc.buildout.easy_install.script_template = getattr(templates, protocol)
            project = self.options.get('projectegg',  self.options['project'])
            
            scripts.extend(
                zc.buildout.easy_install.scripts(
                    [('%s.%s' % (self.options.get('project', self.name), protocol),
                      '', '')],
                    ws,
                    self.options['executable'],
                    self.options['bin-directory'],
                    extra_paths=extra_paths,
                    arguments= "%s.project.%s.settings" % (project,
                                self.options['settings'])))
        zc.buildout.easy_install.script_template = _script_template
        
        self.create_file("src/%s/deploy/%s.wsgi"%(project, project), templates.wsgi, 
                         {'path': 'os.path.join(os.path.abspath(os.path.dirname(__file__)), "..", "%s"),\n  os.path.join(os.path.abspath(os.path.dirname(__file__)), "..", "%s")'%(self.options['local-apps'], self.options['external-apps']),
                          'arguments': "%s.project.%s.settings"%(project,
                                                                 self.options['settings'])})
        
        self.create_file("src/%s/deploy/%s.fcgi"%(project, project), templates.wsgi, 
                         {'path': 'os.path.join(os.path.abspath(os.path.dirname(__file__)), "..", "%s"),\n  os.path.join(os.path.abspath(os.path.dirname(__file__)), "..", "%s")'%(self.options['local-apps'], self.options['external-apps']),
                          'arguments': "%s.project.%s.settings"%(project,
                                                                 self.options['settings'])})
        
        return scripts
    
    def create_project(self, project_dir, project):
        old_config = self.buildout._read_installed_part_options()[0]
        if self.name in old_config:
            old_config = old_config[self.name]
        
            if 'project' in old_config and\
               old_config['project'] != self.options['project']:
                self.log.warning("DjBuild: creating new project '%s', to replace previous project '%s'"%(self.options['project'], old_config['project']))
        
        # saving current work directory
        old_cwd = os.getcwd()
        os.chdir(project_dir)
        
        # importing current django instalation
        import django.core.management
        argv = ['django-admin', 'startproject', project]
        utility = django.core.management.ManagementUtility(argv)
        utility.execute()
        
        self.log.info('DjBuild: creating basic structure')
        os.chdir(project)
        
        os.makedirs('docs')
        os.makedirs('tests')
        os.makedirs('media')
        os.makedirs('deploy')
                
        os.makedirs('static/css')
        os.makedirs('static/js')
        os.makedirs('static/images')
        
        os.makedirs('project/development')
        os.makedirs('project/production')
        os.makedirs('project/testing')
        
        self.create_file('tests/__init__.py', '', {})
        self.create_file('project/__init__.py', '', {})
        self.create_file('project/development/__init__.py', '', {})
        self.create_file('project/production/__init__.py', '', {})
        self.create_file('project/testing/__init__.py', '', {})
        
        self.create_file("project/development/settings.py", templates.development_settings, {'project':self.options['project']})
        self.create_file("project/production/settings.py", templates.production_settings, {'project':self.options['project']})
        self.create_file("project/testing/settings.py", templates.testing_settings, {'project':self.options['project']})
        
        self.create_file("project/development/urls.py", templates.development_urls, {'project':self.options['project']})
        self.create_file("project/production/urls.py", templates.production_urls, {'project':self.options['project']})
        self.create_file("project/testing/urls.py", templates.testing_urls, {'project':self.options['project']})
        
        os.makedirs('templates')
        self.create_file("templates/base.html", templates.base_html)
        self.create_file("templates/404.html", templates.t_404_html)
        self.create_file("templates/500.html", templates.t_500_html)
        
        os.makedirs('logs')
        os.makedirs('logs/development')
        os.makedirs('logs/testing')
        os.makedirs('logs/production')
        
        self.create_file("logs/development/error.log", "")
        self.create_file("logs/testing/error.log", "")
        self.create_file("logs/production/error.log", "")
        
        self.create_file("logs/development/access.log", "")
        self.create_file("logs/testing/access.log", "")
        self.create_file("logs/production/access.log", "")
        
        # updating to original cwd
        os.chdir(old_cwd)
        
    def update_project_structure(self):
        old_config = self.buildout._read_installed_part_options()[0]
        # updating old config to project name
        if self.name in old_config:
            old_config = old_config[self.name]
        
            if 'local-apps' in old_config and\
               old_config['local-apps'] != self.options['local-apps']:
                if os.path.exists(old_config['local-apps']):
                    self.log.info("DjBuild: moving local-apps dir from %s to %s"%(old_config['local-apps'], self.options['local-apps']))
                    shutil.move(old_config['local-apps'], self.options['local-apps'])
                    
            if 'external-apps' in old_config and\
               old_config['external-apps'] != self.options['external-apps']:
                if os.path.exists(old_config['external-apps']):
                    self.log.info("DjBuild: moving external-apps dir from %s to %s"%(old_config['external-apps'], self.options['external-apps']))
                    shutil.move(old_config['external-apps'], self.options['external-apps'])
            
        if not os.path.exists(self.options['local-apps']):
            self.log.info("DjBuild: creating local-apps dir %s"%(self.options['local-apps']))
            os.makedirs(self.options['local-apps'])
            
        if not os.path.exists(self.options['external-apps']):
            self.log.info("DjBuild: creating external-apps dir %s"%(self.options['external-apps']))
            os.makedirs(self.options['external-apps'])
            
        answer = raw_input("Do you want to install/update apps?(yes/no): ")
        
        if answer.lower() == 'yes':
            print '\n************** Intalling django apps **************\n'
            apps = self.options.get('apps', '').split()
            if len(apps) == 0:
                self.log.info('No apps to install')
            else:
                install_dir = os.path.abspath(self.options['external-apps'])
                args = ['-U', '-b', self.buildout['buildout']['download-cache'], '-d', install_dir]
                links = self.options.get('find-links', '').split()
                
                if len(links)>0:
                    links.insert(0, '-f')
                    args.extend(links)
                    
                args.extend(apps)
                install(args)
            print '\n************** Intalling django apps **************\n'
            
        print'\n************** Searching for apps config **************\n'
        
        os.environ["DJANGO_SETTINGS_MODULE"] = '%s.project.%s.settings' % (self.options['project'], self.options['settings'])
               
        from django.conf import settings
        
        apps = os.listdir(self.options['external-apps'])
        local_apps = os.listdir(self.options['local-apps'])
        
        # apps into external apps 
        dne = [ x for x in settings.INSTALLED_APPS[:] if x.find('.') == -1 and x not in local_apps] # django and not external
        end = [] # external and not django
        
        del local_apps
        
        for app in apps:
            if app not in settings.INSTALLED_APPS:
                end.append(app)
            elif app in dne:
                dne.remove(app)
        
        # searching for installed apps into djando but not into external-apps directory
       
        if len(end) > 0:
            self.log.warning('These sources apps are installed but not used into django project: \n\t* %s\n'%'\n\t* '.join(end))
                
        # searching for non installed apps into djando but installed into external-apps directory
        if len(dne) > 0:
            self.log.warning('These apps are installed into django but source not installed: \n\t* %s'%'\n\t* '.join(dne))
            
        if len(dne) == 0 and len(end) == 0:
            self.log.info('\nAll is synced')
        
        print '\n************** Searching for apps config **************\n'
        
        del os.environ["DJANGO_SETTINGS_MODULE"]
    
    def verify_or_create_download_dir(self, download_dir):
        if not os.path.exists(download_dir):
            os.mkdir(download_dir)
            
    def remove_pre_existing_installation(self, location):
        # Remove a pre-existing installation if it is there
        if os.path.exists(location):
            shutil.rmtree(location)
            
    def install_django(self, version, download_dir, location, install_from_cache):
        path = ''
        if self.is_svn_url(version):
            path = self.install_svn_version(version, download_dir, location,
                                            install_from_cache)
        else:
            tarball = self.get_release(version, download_dir)
            # Extract and put the dir in its proper place
            path = self.install_release(version, download_dir, tarball, location)
        # updating path to import django
        import sys
        sys.path.insert(0, path)
            
    def install_recipe(self, location):
        self.options['setup'] = location
        development = zc.recipe.egg.Develop(self.buildout,
                                            self.options['recipe'],
                                            self.options)
        development.install()
        del self.options['setup']
        
    def install_project(self, project_dir, project):
        if not os.path.exists('src'):
            print "creating src directory, it not exists"
            os.mkdir('src')
        if not self.options.get('projectegg'):
            if not os.path.exists(os.path.join(project_dir, project)):
                self.create_project(project_dir, project)
            else:
                self.log.info(
                    'Skipping creating of project: %(project)s since '
                    'it exists' % self.options)
                
            # creating structure
            old_cwd = os.getcwd()
            os.chdir(os.path.join(project_dir, project))
            
            self.update_project_structure()
            
            os.chdir(old_cwd)
                
    def install_scripts(self, extra_path, ws):
        script_paths = []

        # Create the Django management script
        script_paths.extend(self.create_manage_script(extra_path, ws))

        ## Create the test runner
        script_paths.extend(self.create_test_runner(extra_path, ws))

        ## Make the wsgi and fastcgi scripts if enabled
        script_paths.extend(self.make_scripts(extra_path, ws))
        
        return script_paths
